lib/dotcom/system_status/groups.ex

```diff
@@ -0,0 +1,465 @@
+    |> then(&[&1])
+  end
+
+  defp mattapan_branches_with_statuses(alerts, time) do
+    "Mattapan"
+    |> alerts_for_route(alerts)
+    |> case do
+      [] ->
+        []
+
+      mattapan_alerts ->
+        mattapan_statuses = mattapan_alerts |> alerts_to_statuses(time)
+
+        [branch_with_statuses_entry(mattapan_statuses, ["Mattapan"])]
+    end
+  end
+
+  # Exchanges a route_id (a line_id or a branch_id - anything that
+  # might correspond to an alert) for a map with that route_id and the
+  # statuses affecting that route.
+  defp add_statuses_for_route(route_id, alerts, time) do
+    %{
+      route_id: route_id,
+      statuses: statuses_for_route(route_id, alerts, time)
+    }
+  end
+
+  defp statuses_for_route(route_id, alerts, time) do
+    route_id
+    |> alerts_for_route(alerts)
+    |> alerts_to_statuses(time)
+  end
+
+  # Returns a branch_with_status entry, to be used in the
+  # branches_with_statuses field in groups/2. If no branch_ids are
+  # provided, then uses an empty array.
+  # 
+  defp branch_with_statuses_entry(statuses, branch_ids \\ []) do
+    %{
+      branch_ids: branch_ids,
+      statuses: statuses
+    }
+  end
+
+  # Given `alerts` and `route_id`, filters out only the alerts
+  # applicable to the given route, using the alert's "informed
+  # entities".
+  defp alerts_for_route(route_id, alerts) do
+    alerts
+    |> Enum.filter(fn %Alert{informed_entity: informed_entity} ->
+      informed_entity
+      |> Enum.any?(fn
+        %{route: ^route_id} -> true
+        %{} -> false
+      end)
+    end)
+  end
+
+  # Maps a list of alerts to a list of statuses that are formatted
+  # according to the system status specifications:
+  # - Identical alerts are grouped together and pluralized.
+  # - Times are given as a kitchen-formatted string, nil, or "Now".
+  # - Statuses are sorted alphabetically.
+  defp alerts_to_statuses(alerts, time) do
+    alerts
+    |> alerts_to_statuses_naive(time)
+    |> consolidate_duplicate_descriptions()
+    |> sort_statuses()
+    |> stringify_times()
+    |> maybe_add_now_text()
+  end
+
+  # Naively maps a list of alerts to a list of statuses, where a
+  # status is a simple structure with a route, a description, and a
+  # few additional fields that determine how it will render in the
+  # frontend.
+  defp alerts_to_statuses_naive(alerts, time)
+
+  # If there are no alerts, then we want a single status indicating
+  # "Normal Service".
+  defp alerts_to_statuses_naive([], _time) do
+    [%{description: "Normal Service", time: nil}]
+  end
+
+  # If there are alerts, then create a starting list of statuses that
+  # maps one-to-one with the alerts provided.
+  defp alerts_to_statuses_naive(alerts, time) do
+    alerts
+    |> Enum.map(fn alert ->
+      alert_to_status(alert, time)
+    end)
+  end
+
+  # Translates an alert to a status:
+  #  - The effect is humanized into a description for the status.
+  #  - If the alert's already active, `time` is set to `nil`.
```

Comment:
I'm not clear on why setting the time to nil means that it is active.

Reply:
The way this will be rendered in the frontend will be something like:
```
case time do
  nil -> description
  _ -> "#{time}: #{description}"
end
```

Thinking about other ways of representing this.... maybe `prefix` instead of `time`, since in a certain sense, that's actually what it is?

Reply:
This code is gone now - the "current" => nil => "no prefix" logic has been moved into `DotcomWeb` and no longer lives here.